### PR TITLE
Copied over some missing Graphics functions from Phalanger.

### PR DESCRIPTION
**Pull request by gordon-matt**
Managed to convert some of the easy functions, but others are commented out, as they are a little too complex for me to convert from System.Drawing to ImageSharp... but I need them all implemented for issue #190 . NOTE: Nothing has been tested!

### CODE REVIEW

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 1]
```diff
@@ -1169,5 +1159,831 @@ static bool imagesave(Context ctx, PhpResource im, PhpValue to/* = null*/, Actio
...
+        #endregion imagesave
+
+        // NEW FUNCTIONS
+
+        #region imagefilter
+
+        /// <summary>
+        /// Applies a filter to an image
+        /// </summary>
+        public static bool imagefilter(PhpResource src_im, int filtertype) => imagefilter(src_im, filtertype, -1, -1, -1, -1);
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: unnecessary to declare so many overrides,. Use default values instead., This slows down runtime significantly.

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 2]
```diff
@@ -63,22 +64,27 @@ public enum ImgType
             /// Used as a return value by <see cref="imagetypes"/>.
             /// </summary>
             GIF = 1,
+
             /// <summary>
             /// Used as a return value by <see cref="imagetypes"/>.
             /// </summary>
             JPG = JPEG,
+
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: not providing an easy survey with so many formatting changes, please avoid that

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 3]
```diff
@@ -102,7 +108,7 @@ public enum ImgType
         public const int IMG_WBMP = (int)ImgType.WBMP;
         public const int IMG_XPM = (int)ImgType.XPM;
 
-        #endregion
+        #endregion ImgType
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: is it necessary?

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 4]
```diff
@@ -346,12 +363,16 @@ public static int imagetypes()
...
+
+            if (dst_w == 0 || dst_h == 0)
+                return true;
+
             //if (dst_w < 0) dst_w = 0;
             //if (dst_h < 0) dst_h = 0;
-            if (dst_w == 0 || dst_h == 0) return true;
 
-            var src = new Image<Rgba32>(src_img.Image)  // copy image and resize inplace:
+            var src = src_img.Image
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: bug bug bug

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 5]
```diff
@@ -334,7 +351,7 @@ public static int imagetypes()
                 new NearestNeighborResampler());
         }
 
-        static bool ImageCopyAndResize(PhpResource dst_im, PhpResource src_im,
+        private static bool ImageCopyAndResize(PhpResource dst_im, PhpResource src_im,
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: we do not write `private` everywhere, also it makes the PR hard to maintain

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 6]
```diff
@@ -695,68 +720,11 @@ public static long imagecolorat(PhpResource im, int x, int y)
             return (long)image[x, y].Rgba;
         }
 
-        /// <summary>
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: keep me

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 7]
```diff
@@ -695,68 +720,11 @@ public static long imagecolorat(PhpResource im, int x, int y)
...
-            PhpException.FunctionNotSupported("imageinterlace");
-            return 0; // false
-        }
-
-        #region imagefilter
-
-        /// <summary>
-        /// Applies a filter to an image.
-        /// </summary>
-        public static bool imagefilter(PhpResource image, FilterTypes filtertype, int arg1 = 0, int arg2 = 0, int arg3 = 0, int arg4 = 0)
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: keep me

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 8]
```diff
@@ -1169,5 +1159,831 @@ static bool imagesave(Context ctx, PhpResource im, PhpValue to/* = null*/, Actio
...
+
+        /// <summary>
+        /// Applies a filter to an image
+        /// </summary>
+        public static bool imagefilter(PhpResource src_im, int filtertype, int arg1, int arg2, int arg3) => imagefilter(src_im, filtertype, arg1, arg2, arg3, -1);
+
+        /// <summary>
+        /// Applies a filter to an image
+        /// </summary>
+        public static bool imagefilter(PhpResource im, int filtertype, int arg1, int arg2, int arg3, int arg4)
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: already implemented

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 9]
```diff
@@ -1169,5 +1159,831 @@ static bool imagesave(Context ctx, PhpResource im, PhpValue to/* = null*/, Actio
...
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Makes specified Bitmap grayscaled
+        /// </summary>
+        /// <param name="bitmap">processed bitmap</param>
+        /// <returns>success indication</returns>
+        private static bool MakeGrayscale(Image<Rgba32>/*!*/bitmap)
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: unnecessary

File: src/Peachpie.Library.Graphics/PhpGd2.cs [hunk 10]
```diff
@@ -1169,5 +1159,831 @@ static bool imagesave(Context ctx, PhpResource im, PhpValue to/* = null*/, Actio
...
+        }
+
+        #endregion imagecolorresolvealpha
+
+        /// <summary>
+        /// Inverts colors in specified Bitmap
+        /// </summary>
+        /// <param name="bitmap">processed bitmap</param>
+        /// <returns>success indication</returns>
+        private static bool InvertColors(Image<Rgba32>/*!*/bitmap)
```
Comment (jakubmisek) [inline src/Peachpie.Library.Graphics/PhpGd2.cs]: unnecessary

**Comment by jakubmisek** [comment]
I cannot merge the PR as it is, but it is on a good way!

May I ask to update from master
- do not make so many changes out of the scope of your added features
- do not use overloads instead of default parameters
- use `PhpException.FunctionNotSupported` for functions that are not supported
- use `throw new NotImplementedException` for functions that are not implemented but will be
- avoid copy/paste from other repos, comments from other repos in Peachpie doesn't seem right

Thanks : )

**Comment by gordon-matt** [comment]
Thanks for the advice. I will see what I can do about changes. By the way, I think most of the changes outside the scope of my added features are from an VS extension I have installed (CodeMaid). I have a habit of running it with a keyboard shortcut every time before I save. You can see it's nothing but removing whitespace really - and I think it might have added some text to the `#endregion` sections.. but yeah, I will be careful not to run that next time.

**Comment by jakubmisek** [comment]
I see, yes it looks like it was generated automatically.
We should probably define a common settings file in root of the project.

**Comment by gordon-matt** [comment]
Changes made, as requested.

**Comment by jakubmisek** [comment]
Thank you! looks great

**Pull request merged**